eZ/Publish/Core/Search/Legacy/Content/IndexerGateway.php

```diff
@@ -0,0 +1,147 @@
+        );
+
+        return (int)$query->execute()->fetchOne();
+    }
+
+    public function getAllContent(int $iterationCount): Generator
+    {
+        $query = $this->buildQueryForAllContent();
+
+        yield from $this->fetchIteration($query->execute(), $iterationCount);
+    }
+
+    public function countAllContent(): int
+    {
+        $query = $this->buildCountingQuery(
+            $this->buildQueryForAllContent()
+        );
+
+        return (int)$query->execute()->fetchOne();
+    }
+
+    private function buildQueryForContentSince(DateTimeInterface $since): QueryBuilder
+    {
+        return $this->connection->createQueryBuilder()
+            ->select('c.id')
+            ->from('ezcontentobject', 'c')
+            ->where('c.status = :status')->andWhere('c.modified >= :since')
+            ->setParameter('status', ContentInfo::STATUS_PUBLISHED, ParameterType::INTEGER)
+            ->setParameter('since', $since->getTimestamp(), ParameterType::INTEGER);
+    }
+
+    private function buildQueryForContentInSubtree(string $locationPath): QueryBuilder
+    {
+        return $this->connection->createQueryBuilder()
+            ->select('DISTINCT c.id')
+            ->from('ezcontentobject', 'c')
+            ->innerJoin('c', 'ezcontentobject_tree', 't', 't.contentobject_id = c.id')
+            ->where('c.status = :status')
+            ->andWhere('t.path_string LIKE :path')
+            ->setParameter('status', ContentInfo::STATUS_PUBLISHED, ParameterType::INTEGER)
+            ->setParameter('path', $locationPath . '%', ParameterType::STRING);
+    }
+
+    private function buildQueryForAllContent(): QueryBuilder
+    {
+        return $this->connection->createQueryBuilder()
+            ->select('c.id')
+            ->from('ezcontentobject', 'c')
+            ->where('c.status = :status')
+            ->setParameter('status', ContentInfo::STATUS_PUBLISHED, ParameterType::INTEGER);
+    }
+
+    /**
+     * @throws \Doctrine\DBAL\Exception
+     */
+    private function buildCountingQuery(QueryBuilder $query): QueryBuilder
+    {
+        $databasePlatform = $this->connection->getDatabasePlatform();
+
+        // wrap existing select part in count expression
+        $query->select(
+            $databasePlatform->getCountExpression(
+                $query->getQueryPart('select')[0]
+            )
+        );
+
+        return $query;
+    }
+
+    private function fetchIteration(ResultStatement $stmt, int $iterationCount): Generator
+    {
+        do {
+            $contentIds = [];
+            for ($i = 0; $i < $iterationCount; ++$i) {
+                if ($contentId = $stmt->fetchOne()) {
+                    $contentIds[] = $contentId;
+                } elseif (empty($contentIds)) {
+                    return;
+                } else {
+                    break;
+                }
+            }
+
+            yield $contentIds;
+        } while (!empty($contentId));
```

Comment:
I have feeling that something is wrong here but maybe it's lack of coffee...

Reply:
Lack of coffee :smile: 

I was also bewildered by this code though, and for this reason wanted phpdoc to include the fact that the generator built here will actually return arrays of integers (`int[]`), not single integers (`int`). But PHPStorm refused to understand the `@return \Generator<int[]>` annotation :stuck_out_tongue_closed_eyes:

Reply:
> I have feeling that something is wrong here but maybe it's lack of coffee...

It's a multidimensional array - a list of Content IDs for each iteration.

Reply:
Variable names are misleading, `$contentId` as an array of IDs brings a lot of confusion to this code.

Reply:
@Nattfarinn actually no. `$contentId` is a single ID indeed.

To resolve the readability issue, I suggest we move the internal iteration to a separate, private method?